lib/transaction.go

```diff
@@ -108,11 +108,51 @@ func (tx Transaction) IsWellFormed(networkID []byte) (err error) {
 	return
 }
 
+// Validate checks,
+// * source account exists
+// * checkpoint is valid
+// * source has enough balance to pay
+// * and it's `Operations`
 func (tx Transaction) Validate(st *sebakstorage.LevelDBBackend) (err error) {
-	// TODO check whether `Checkpoint` is in `Block Transaction` and is latest
-	// `Checkpoint`
-	// TODO check whether `Source` is in `Block Account`
-	// TODO check whether the balance of `Source` is greater than `totalAmount`
+	// check, source exists
+	var ba *BlockAccount
+	if ba, err = GetBlockAccount(st, tx.B.Source); err != nil {
+		err = sebakerror.ErrorBlockAccountDoesNotExists
+		return
+	}
+
+	// check, checkpoint is based on latest checkpoint
+	if !tx.IsValidCheckpoint(ba.Checkpoint) {
+		err = sebakerror.ErrorTransactionInvalidCheckpoint
+		return
+	}
+
+	// get the balance at checkpoint
+	var bac BlockAccountCheckpoint
+	bac, err = GetBlockAccountCheckpoint(st, tx.B.Source, tx.B.Checkpoint)
+	if err != nil {
+		return
+	}
+
+	totalAmount := tx.TotalAmount(true)
+
+	// check, have enough balance at checkpoint
+	if sebakcommon.MustAmountFromString(bac.Balance) < totalAmount {
+		err = sebakerror.ErrorTransactionExcessAbilityToPay
+		return
+	}
+
+	// check, have enough balance now
+	if sebakcommon.MustAmountFromString(ba.Balance) < totalAmount {
+		err = sebakerror.ErrorTransactionExcessAbilityToPay
+		return
+	}
+
+	for _, op := range tx.B.Operations {
```

Comment:
I would say move this before the balance check, because if any operation is invalid `TotalAmount` does not make sense / could error out.

Reply:
@Geod24 We can choose the proper position of this block. Anyway `Operation.Validate` is more expensive than checking `TotalAmount`.